hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/spark/sql/execution/benchmark/CowTableReadBenchmark.scala

```diff
@@ -0,0 +1,128 @@
+
+  def getSparkSession: SparkSession = SparkSession.builder()
+    .master("local[4]")
+    .appName(this.getClass.getCanonicalName)
+    .withExtensions(new HoodieSparkSessionExtension)
+    .config("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
+    .config("hoodie.insert.shuffle.parallelism", "2")
+    .config("hoodie.upsert.shuffle.parallelism", "2")
+    .config("hoodie.delete.shuffle.parallelism", "2")
+    .config("spark.sql.session.timeZone", "CTT")
+    .config(sparkConf())
+    .getOrCreate()
+
+  def sparkConf(): SparkConf = {
+    val sparkConf = new SparkConf()
+    if (HoodieSparkUtils.gteqSpark3_2) {
+      sparkConf.set("spark.sql.catalog.spark_catalog",
+        "org.apache.spark.sql.hudi.catalog.HoodieCatalog")
+    }
+    sparkConf
+  }
+
+  def prepareHoodieCowTable(tableName: String, tablePath: String) = {
+    createDataFrame(10000000).registerTempTable("ds")
+    spark.sql(
+      s"""
+         |create table $tableName using hudi
+         |tblproperties(primaryKey = 'c1')
+         |location '${tablePath}'
+         |As
+         |select * from ds
+       """.stripMargin)
+  }
+
+  private def createDataFrame(number: Int): DataFrame = {
+    val schema = new StructType()
+      .add("c1", IntegerType)
+      .add("c11", IntegerType)
+      .add("c12", IntegerType)
+      .add("c2", StringType)
+      .add("c3", DecimalType(38, 10))
+      .add("c4", TimestampType)
+      .add("c5", ShortType)
+      .add("c6", DateType)
+      .add("c7", BinaryType)
+      .add("c9", ByteType)
+
+    val rdd = spark.sparkContext.parallelize(0 to number, 2).map { item =>
+      val c1 = Integer.valueOf(item)
+      val c11 = Integer.valueOf(Random.nextInt(10000))
+      val c12 = Integer.valueOf(Random.nextInt(10000))
+      val c2 = s" ${item}abc"
+      val c3 = new java.math.BigDecimal(s"${Random.nextInt(1000)}.${Random.nextInt(100)}")
+      val c4 = new Timestamp(System.currentTimeMillis())
+      val c5 = java.lang.Short.valueOf(s"${16}")
+      val c6 = Date.valueOf(s"${2020}-${item % 11 + 1}-${item % 28 + 1}")
+      val c7 = Array(item).map(_.toByte)
+      val c8 = java.lang.Byte.valueOf("9")
+      RowFactory.create(c1, c11, c12, c2, c3, c4, c5, c6, c7, c8)
+    }
+    spark.createDataFrame(rdd, schema)
+  }
+
+  def withTempTable(tableNames: String*)(f: => Unit): Unit = {
+    try f finally tableNames.foreach(spark.catalog.dropTempView)
+  }
+
+  private def cowTableReadBenchmark(tableName: String = "cowBenchmark"): Unit = {
+    withTempDir {f =>
+      withTempTable(tableName) {
+        prepareHoodieCowTable(tableName, new Path(f.getCanonicalPath, tableName).toUri.toString)
+        val benchmark = new HoodieBenchmark("perf cow snapshot read", 10000000)
+        benchmark.addCase("vectorized disable") { _ =>
+          spark.sessionState.conf.setConfString("spark.sql.parquet.enableVectorizedReader", "false")
+          spark.sql(s"select c1, c3, c4, c5 from $tableName").count()
+        }
```

Comment:
Maybe we can past the current benchmark result into the code comments so that we can diff with it for following releases.

Reply:
yes, of cource, let me do it